sql/src/main/java/io/crate/operation/scalar/cast/AbstractCastFunction.java

```diff
@@ -80,22 +83,32 @@ public Symbol normalizeSymbol(Function symbol) {
             Object value = ((Input) argument).value();
             try {
                 return Literal.newLiteral(returnType, returnType.value(value));
-            } catch (ClassCastException | IllegalArgumentException | ConversionException e) {
-                return onNormalizeException(argument, e);
+            } catch (ClassCastException | IllegalArgumentException | ConversionException | InvalidIPException e) {
+                return onNormalizeException(argument);
             }
         }
         return symbol;
     }
 
-    protected Symbol onNormalizeException(Symbol argument, Throwable t) {
-        throw new IllegalArgumentException(
-                String.format(Locale.ENGLISH, "cannot cast %s to %s",
-                        SymbolPrinter.INSTANCE.printSimple(argument), returnType), t);
+    protected Symbol onNormalizeException(Symbol argument) {
+        throw new ConversionException(
+                String.format(Locale.ENGLISH, "cannot cast %s to type %s",
+                        SymbolPrinter.INSTANCE.printSimple(argument), returnType));
     }
 
-    protected To onEvaluateException(From argument, Throwable t) {
-        Throwables.propagate(t);
-        return null;
+    protected To onEvaluateException(From argument) {
+        String value;
+        if (argument instanceof BytesRef) {
+            value = String.format(Locale.ENGLISH, "'%s'", ((BytesRef) argument).utf8ToString());
+        } else if (argument instanceof String) {
+            value = String.format(Locale.ENGLISH, "'%s'", argument);
+        } else {
+            value = argument.toString();
+        }
+        throw new ConversionException(
+                String.format(Locale.ENGLISH, "cannot cast %s to type %s",
```

Comment:
would it make sense to change the ConversionException to take the value and returnType and do the formatting in the ctor to avoid the formatting handling in different places?